In [1]:
def recur_fibo(n):
    if n <= 1:
        return n
    else:
        return recur_fibo(n - 1) + recur_fibo(n - 2)

nterms = 20

if nterms <= 0:
    print("Please enter a positive integer")
else:
    print("Fibonacci sequence:")
    for i in range(nterms):
        print(recur_fibo(i))

Fibonacci sequence:
0
1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181


In [2]:
def recur_fibo(n):
    if n <= 1:
        return n
    else:
        return recur_fibo(n - 1) + recur_fibo(n - 2)

# Menerima input dari pengguna
nterms = int(input("Masukkan jumlah angka Fibonacci yang diinginkan: "))

if nterms <= 0:
    print("Silakan masukkan angka positif.")
else:
    print(f"Deret Fibonacci sebanyak {nterms} angka:")
    for i in range(nterms):
        print(recur_fibo(i))


Masukkan jumlah angka Fibonacci yang diinginkan: 30
Deret Fibonacci sebanyak 30 angka:
0
1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765
10946
17711
28657
46368
75025
121393
196418
317811
514229


In [5]:
import heapq

# Definisikan kelas Node untuk pohon Huffman
class Node:
    def __init__(self, freq, symbol, left=None, right=None):
        """
        Inisialisasi node Huffman dengan:
        - freq: frekuensi kemunculan simbol
        - symbol: simbol karakter (misalnya 'a', 'b', dsb.)
        - left dan right: anak kiri dan kanan (untuk node internal)
        """
        self.freq = freq        # Frekuensi simbol
        self.symbol = symbol    # Simbol karakter
        self.left = left        # Anak kiri
        self.right = right      # Anak kanan
        self.huff = ''          # Kode Huffman yang akan diisi saat pohon dibangun

    # Membandingkan node berdasarkan frekuensi agar bisa digunakan dalam heap
    def __lt__(self, nxt):
        return self.freq < nxt.freq

# Fungsi untuk mencetak kode Huffman dari pohon
def printNodes(node, val=''):
    """
    Fungsi rekursif untuk mencetak kode Huffman.
    Dimulai dari root, kode '0' ditambahkan jika ke kiri, dan '1' jika ke kanan.
    """
    # Menambahkan kode saat ini (0 atau 1) ke jalur yang ada
    newVal = val + str(node.huff)

    # Jika node memiliki anak kiri, lanjutkan ke subtree kiri
    if node.left:
        printNodes(node.left, newVal)

    # Jika node memiliki anak kanan, lanjutkan ke subtree kanan
    if node.right:
        printNodes(node.right, newVal)

    # Jika node adalah daun (tidak memiliki anak kiri/kanan), cetak simbol dan kodenya
    if not node.left and not node.right:
        print(f"{node.symbol} -> {newVal}")

# Daftar karakter dan frekuensi yang sesuai
chars = ['a', 'b', 'c', 'd', 'e', 'f']
freq = [5, 9, 12, 13, 16, 45]

# Inisialisasi heap (priority queue) dari node
# Setiap simbol dan frekuensinya dimasukkan ke dalam heap sebagai node daun
nodes = []
for x in range(len(chars)):
    # Membuat node baru untuk setiap simbol dan frekuensi, lalu memasukkannya ke dalam heap
    heapq.heappush(nodes, Node(freq[x], chars[x]))

# Bangun pohon Huffman
# Proses utama algoritma, dilakukan sampai hanya ada satu node (root) tersisa
while len(nodes) > 1:
    # Mengambil dua node dengan frekuensi terkecil dari heap
    left = heapq.heappop(nodes)
    right = heapq.heappop(nodes)

    # Tetapkan kode '0' ke node kiri dan '1' ke node kanan
    left.huff = '0'
    right.huff = '1'

    # Buat node internal baru dengan frekuensi total dari dua node yang diambil
    newNode = Node(left.freq + right.freq, left.symbol + right.symbol, left, right)

    # Masukkan node internal baru ini kembali ke dalam heap
    heapq.heappush(nodes, newNode)

# Cetak hasil kode Huffman dari pohon yang sudah dibangun
# Menelusuri pohon dari root hingga ke daun, mencetak kode untuk setiap simbol
printNodes(nodes[0])


f -> 0
c -> 100
d -> 101
a -> 1100
b -> 1101
e -> 111


Tahapan Pembuatan Huffman Coding Tree
1.	Inisialisasi Node: Pertama-tama, kita buat node daun untuk setiap simbol beserta frekuensinya. Contohnya, untuk simbol 'a' yang memiliki frekuensi 5, kita buat satu node.
2.	Membuat Min-Heap: Setelah itu, semua node daun dimasukkan ke dalam struktur data min-heap. Di sini, node dengan frekuensi terendah akan selalu diambil terlebih dahulu.
3.	Penggabungan Node: Selanjutnya, ambil dua node yang memiliki frekuensi terkecil. Node pertama kita beri kode '0' sebagai anak kiri, dan node kedua kita beri kode '1' sebagai anak kanan. Kemudian, kita buat node baru yang menjumlahkan frekuensi kedua node tersebut dan masukkan kembali ke dalam heap.
4.	Ulangi Proses: Proses penggabungan ini diulang terus-menerus hingga hanya tersisa satu node, yang akan menjadi root dari pohon Huffman.
5.	Pencetakan Kode: Terakhir, kita telusuri pohon dari root untuk mendapatkan kode Huffman untuk setiap simbol. Misalnya, kita bisa mendapatkan kode seperti a -> 1100, b -> 1101, dan seterusnya.

